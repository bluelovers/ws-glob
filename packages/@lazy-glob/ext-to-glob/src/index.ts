import { escapeRegExp } from 'regexp-helper-core';


export type IExtString<T extends string = string> = `.${T}`

export function removeDot<T extends string>(ext: IExtString<T>)
{
	return ext.replace(/^\./, '') as T
}

export function removeDotFromExtensions<T extends string>(arr: readonly IExtString<T>[], fns?: ((input: string) => string)[])
{
	return arr.map(input => {
		input = removeDot(input) as any;
		fns?.forEach(fn => {
			input = fn(input) as any;
		})
		return input
	})
}

export function extToGlob<T extends string>(arr: readonly IExtString<T>[])
{
	return `.+(${removeDotFromExtensions(arr).join('|')})` as const
}

export function extToRegexpPattern<T extends string>(arr: readonly IExtString<T>[])
{
	return `.(${removeDotFromExtensions(arr, [escapeRegExp]).join('|')})$` as const
}

export default extToGlob
