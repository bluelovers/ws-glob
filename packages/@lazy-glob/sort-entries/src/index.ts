/**
 * Created by user on 2020/6/9.
 */

import { naturalCompare } from '@bluelovers/string-natural-compare';
import { ITree } from '@lazy-glob/util/lib/types';
import { ISortEntriesSort, ISortEntriesSortRow } from '@lazy-glob/util/lib/types/entries';
import { SymGlobTree } from '@lazy-glob/util';

type IEntry<T extends Record<any, any>, K extends keyof T> = [K, T[K]]
type IEntries<T extends ITree = ITree> = IEntry<T, keyof T>[];

export type IEntriesInput = IEntries | ISortEntriesSort;

export function entries_sort(entries: IEntriesInput, fn: (a: string, b: string, cache: any) => number = naturalCompare, cache: any = {}): ISortEntriesSort
{
	entries = (entries as ISortEntriesSort).reduce(function (a: ISortEntriesSort, b: ISortEntriesSortRow)
	{
    const v = b[1];

    if (v === null || typeof v == 'string')
		{
			a.push(b);
		}
		else
		{
      const d = Object.entries(v);
      a.push([b[0], entries_sort(d, fn), (v as any)[SymGlobTree]]);
		}

		return a;
	}, [] as ISortEntriesSort);

	(entries as ISortEntriesSort).sort(function (a, b)
	{
    const r = fn(a[0], b[0], cache);
    return r;
	});

	return entries;
}

export function entries_reduce(entries: IEntriesInput)
{
	return (entries as ISortEntriesSort)
		.reduce(function (a, [k, v, bool]: ISortEntriesSortRow)
		{
			if (v === null || typeof v == 'string')
			{
				a[k] = v as string;
			}
			else
			{
				a[k] = entries_reduce(v);
			}

			if (bool)
			{
        // @ts-ignore
				a[k][SymGlobTree] = bool;
			}

			return a;
		}, {} as ITree)
		;
}

export function sort(a: ITree, fn: (a: string, b: string, cache: any) => number = naturalCompare)
{
  const r = Object.entries(a);

  const a1 = entries_sort(r, fn);
  const a2 = entries_reduce(a1);

  return a2
}

export default sort
