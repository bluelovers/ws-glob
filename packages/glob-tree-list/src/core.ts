/**
 * Created by user on 2018/3/29/029.
 */

//export * from './lib/types'
import { SymGlobTree } from '@lazy-glob/util';

export * from '@lazy-glob/util/lib/types';
import path from 'upath2';

import { ITree } from '@lazy-glob/util/lib/types';
export { path }

export function globToTree(data: string[]): ITree
{
	return data.reduce(function (a, b)
	{
		b = path.normalize(b);
		let dirname = path.dirname(b);
		let basename = path.basename(b);

		let isdir = b.slice(-1) == path.sep;

		if (isdir)
		{
			basename += path.sep;
		}

		//console.log([dirname, basename]);

		if (dirname == '.')
		{
			let f = a;

			f[basename] = isdir ? null : basename;
		}
		else
		{
			let c = dirname
				.split(path.sep)
			;

			if (c[0] == '.')
			{
				c.shift();
			}

			let f = a;

			c.forEach(function (e)
			{
				e += path.sep;

        // @ts-ignore
				f[e] = f[e] || {};

				// @ts-ignore
				f = f[e];
			});

			f[basename] = isdir ? (f[basename] || {}) : basename;

			if (isdir)
			{
        // @ts-ignore
				f[basename][SymGlobTree] = true;
				//console.dir({ b, basename, f })
			}
		}

		return a;
	}, {} as ITree);
}

export function treeToGlob(a: ITree, d: string[] = []): string[]
{
	return Object.entries(a).reduce(function (a, b)
	{
		//console.log(b);

		if (b[1] === null || typeof b[1] == 'string')
		{
			let k = (b[1] === null ? b[0] : b[1]) as string;

			if (d.length)
			{
				// @ts-ignore
				a.push(path.join(...d, k));
			}
			else
			{
				a.push(k);
			}
		}
		else
		{
			let ls = treeToGlob(b[1], d.concat(b[0]));

			if (b[1][SymGlobTree])
			{
				let k = b[0];

				if (d.length)
				{
					// @ts-ignore
					a.push(path.join(...d, k));
				}
				else
				{
					a.push(k);
				}
			}

			a = a.concat(ls);
		}

		return a;
	}, [] as string[]);
}

globToTree.globToTree = globToTree;
globToTree.treeToGlob = treeToGlob;
globToTree.default = globToTree;

export default exports as typeof import('./core');

