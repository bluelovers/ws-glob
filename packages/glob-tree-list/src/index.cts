
import { globToTree } from './index';

// @ts-ignore
export = globToTree
