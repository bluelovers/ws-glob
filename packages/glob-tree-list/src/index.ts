/**
 * Created by user on 2018/3/29/029.
 */

import { SymGlobTree } from '@lazy-glob/util';
import {
	join,
	normalize,
	sep,
	dirname as _dirname,
	basename as _basename,
} from 'upath2';
import { ITree } from '@lazy-glob/util/lib/types';

export type { ITree }

export function globToTree(data: string[]): ITree
{
	return data.reduce(function (a, b)
	{
		b = normalize(b);
		let dirname = _dirname(b);
		let basename = _basename(b);

		const isdir = b.slice(-1) === sep;

		if (isdir)
		{
			basename += sep;
		}

		//console.log([dirname, basename]);

		if (dirname === '.')
		{
			const f = a;

			f[basename] = isdir ? null : basename;
		}
		else
		{
			const c = dirname
				.split(sep)
			;

			if (c[0] === '.')
			{
				c.shift();
			}

			let f = a;

			c.forEach(function (e)
			{
				e += sep;

				// @ts-ignore
				f[e] = f[e] || {};

				// @ts-ignore
				f = f[e];
			});

			f[basename] = isdir ? (f[basename] || {}) : basename;

			if (isdir)
			{
				// @ts-ignore
				f[basename][SymGlobTree] = true;
				//console.dir({ b, basename, f })
			}
		}

		return a;
	}, {} as ITree);
}

export function treeToGlob(a: ITree, d: string[] = []): string[]
{
	return Object.entries(a).reduce(function (a, b)
	{
		//console.log(b);

		if (b[1] === null || typeof b[1] === 'string')
		{
			const k = (b[1] === null ? b[0] : b[1]) as string;

			if (d.length)
			{
				a.push(join(...d as [string], k));
			}
			else
			{
				a.push(k);
			}
		}
		else
		{
			const ls = treeToGlob(b[1], d.concat(b[0]));

			if (b[1][SymGlobTree])
			{
				let k = b[0];

				if (d.length)
				{
					a.push(join(...d as [string], k));
				}
				else
				{
					a.push(k);
				}
			}

			a = a.concat(ls);
		}

		return a;
	}, [] as string[]);
}

Object.defineProperty(globToTree, "__esModule", { value: true });
Object.defineProperty(globToTree, "globToTree", { value: globToTree });
Object.defineProperty(globToTree, "treeToGlob", { value: treeToGlob });
Object.defineProperty(globToTree, "default", { value: globToTree });

export default globToTree
