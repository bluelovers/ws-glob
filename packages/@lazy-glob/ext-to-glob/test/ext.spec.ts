//@noUnusedParameters:false
/// <reference types="jest" />
/// <reference types="node" />
/// <reference types="expect" />

import { basename, extname } from 'path';
import { extToGlob, extToRegexpPattern, removeDotFromExtensions } from '../src/index';

beforeAll(async () =>
{

});

const demo = [
	".cts", ".mts", ".umd.ts", ".ts", ".tsx", ".jsx", ".cjs", ".mjs", ".umd.js", ".js"
] as const;

test(`extToGlob`, () =>
{

	let actual = extToGlob(demo);

	expect(actual).toMatchSnapshot();

});

test(`extToRegexpPattern`, () =>
{

	let actual = extToRegexpPattern(demo);

	expect(actual).toMatchSnapshot();

});

test(`removeDotFromExtensions`, () =>
{

	let actual = removeDotFromExtensions(demo);

	expect(actual).toMatchSnapshot();

});
