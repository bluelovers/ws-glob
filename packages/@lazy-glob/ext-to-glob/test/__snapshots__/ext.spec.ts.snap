// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`extToGlob 1`] = `".+(cts|mts|umd.ts|ts|tsx|jsx|cjs|mjs|umd.js|js)"`;

exports[`extToRegexpPattern 1`] = `".(cts|mts|umd\\.ts|ts|tsx|jsx|cjs|mjs|umd\\.js|js)$"`;

exports[`removeDotFromExtensions 1`] = `
[
  "cts",
  "mts",
  "umd.ts",
  "ts",
  "tsx",
  "jsx",
  "cjs",
  "mjs",
  "umd.js",
  "js",
]
`;
